src/main/java/edu/ucla/library/avpairtree/handlers/LocalstackS3WaveformConsumer.java

```diff
@@ -0,0 +1,220 @@
+     * replies with the URL of the object.
+     *
+     * @param aMessage A message containing the waveform data and headers
+     * @throws IllegalArgumentException If an object key was not supplied in the message headers
+     */
+    @Override
+    public void handle(final Message<byte[]> aMessage) throws IllegalArgumentException {
+        final MultiMap headers = aMessage.headers();
+        final Builder putRequestBuilder = PutObjectRequest.builder().bucket(myS3BucketName);
+        final PutObjectRequest putRequest;
+        final String s3ObjectKey;
+
+        // Required
+        if (headers.contains(KEY)) {
+            s3ObjectKey = headers.get(KEY);
+            putRequestBuilder.key(s3ObjectKey);
+        } else {
+            throw new IllegalArgumentException(LOGGER.getMessage(MessageCodes.AVPT_024));
+        }
+
+        // Optional
+        if (headers.contains(CONTENT_ENCODING)) {
+            putRequestBuilder.contentEncoding(headers.get(CONTENT_ENCODING));
+        }
+
+        putRequest = putRequestBuilder.build();
+
+        myS3Client.putObject(putRequest, AsyncRequestBody.fromBytes(aMessage.body())).whenComplete((resp, err) -> {
+            if (resp != null) {
+                // Success!
+                final String audiowaveformURL = StringUtils.format(myS3ObjectUrlTemplate,
+                        URLEncoder.encode(s3ObjectKey, StandardCharsets.UTF_8));
+
+                // Reply with a JsonObject associating the item ARK with the URL for the audiowaveform data
+                aMessage.reply(audiowaveformURL);
+            } else {
+                final String s3ErrorMsg = LOGGER.getMessage(MessageCodes.AVPT_022, s3ObjectKey, err.getMessage());
+
+                // Since the sender (WatcherVerticle) just logs all errors, should be okay to use a single
+                // failureCode for all errors
+                aMessage.fail(Op.ERROR_CODE, s3ErrorMsg);
+            }
+        });
+    }
+
+    /**
+     * Creates a S3 bucket using the name passed to the constructor.
+     *
+     * @return A Future that will be completed when the bucket creation is complete
+     */
+    public Future<Void> createBucket() {
+        final Promise<Void> promise = Promise.promise();
+        final CreateBucketRequest request = CreateBucketRequest.builder().bucket(myS3BucketName).build();
+
+        myS3Client.createBucket(request).whenComplete((resp, err) -> {
+            if (resp != null) {
+                promise.complete();
+            } else {
+                promise.fail(err);
+            }
+        });
+
+        return promise.future();
+    }
+
+    /**
+     * Deletes all the objects in the S3 bucket whose name was passed to the constructor. This assumes that there are
+     * not more than 1000 objects in the bucket.
+     *
+     * @return A Future that will be completed when the object deletions are complete
+     */
+    public Future<Void> clearBucket() {
```

Comment:
I believe localstack "s3" environments are purely in memory so will be cleaned up automatically when the test ends(?)

Reply:
You're right that it's not strictly necessary to clear the bucket at this time since WaveformVerticleIT has only one test (as you said, it will be "cleaned up" when the localstack container is destroyed), but if more tests are added then it might be necessary. Since more tests may be added in the future, depending on the container termination for test cleanup didn't seem right to me.

Reply:
I would suggest that more tests should use different S3 keys so wouldn't be affected by any other S3 objects.

Reply:
Ok. Should I remove this code?

Reply:
I think so.

Reply:
Ok, done.